libs/platform/src/lib/table/table.service.ts

```diff
@@ -226,9 +263,126 @@ export class TableService {
+        if (!event) {
+            const currCell = this._tableCellsMap.get(currCellNavId);
+
+            if (currCell) {
+                currCell.focus();
+            }
+
+            return;
+        }
+
+        const [rowIndex, colIndex] = currCellNavId.split(',');
+        let nextCellId: string;
+
+        if (KeyUtil.isKeyCode(event, UP_ARROW)) {
+            nextCellId = `${+rowIndex - 1},${colIndex}`;
+        }
+
+        if (KeyUtil.isKeyCode(event, DOWN_ARROW)) {
+            nextCellId = `${+rowIndex + 1},${colIndex}`;
+        }
+
+        if (KeyUtil.isKeyCode(event, LEFT_ARROW)) {
+            nextCellId = `${rowIndex},${+colIndex - (this._rtl ? -1 : 1)}`;
+        }
+
+        if (KeyUtil.isKeyCode(event, RIGHT_ARROW)) {
+            nextCellId = `${rowIndex},${+colIndex + (this._rtl ? -1 : 1)}`;
+        }
+
+        const nextCell = this._tableCellsMap.get(nextCellId);
+        if (nextCell) {
+            this._tableCellsMap.get(nextCellId).focus();
+        }
+    }
+
+    /** Set currently focused cell */
+    setFocusedTableCell(cellId): void {
+        this._focusedTableCellNavId = cellId;
+    }
+
+    /** Reset currently focused cell to initial value */
+    resetFocusedTableCell(): void {
+        this._focusedTableCellNavId = FIRST_CELL_NAVIGATION_ID;
+    }
+
+    /** Mark that focus inside the cell */
+    setFocusInsideCell(): void {
+        this._isFocusInsideTableCell = true;
+    }
+
+    /** Mark that focus isn't inside the cell */
+    removeFocusInsideCell(): void {
+        this._isFocusInsideTableCell = false;
+    }
+
+    /** Focus on table cell inner any first focusable element */
+    processFocusInsideCell(tableCellNavId: TableCellNavigationId): void {
+        const tableCell = this._tableCellsMap.get(tableCellNavId);
+
+        if (this.isFocusInsideTableCell) {
+            tableCell.focus();
+            this.removeFocusInsideCell();
+            return;
+        }
+
+        const focusableElements = getFocusableElements(tableCell._elRef.nativeElement);
+
+        if (focusableElements.length) {
+            focusableElements[0].focus();
+            this.setFocusInsideCell();
+        }
+    }
+
+    /** Process focus when navigating between cell inners */
+    processFocusBetweenCellInners(event: KeyboardEvent, tableContainer: HTMLElement, focusableMock: HTMLElement): void {
+        if (!this.isFocusInsideTableCell) {
+            return;
+        }
+
+        const focusableElements = getFocusableElements(tableContainer);
+        const currFocusElementIndex = focusableElements.findIndex((el) => el === document.activeElement);
+
+        if (currFocusElementIndex === 1 && event.shiftKey) {
+            this.removeFocusInsideCell();
+        }
+
+        if (currFocusElementIndex === focusableElements.length - 1 && !event.shiftKey) {
+            focusableMock.focus();
+            this.removeFocusInsideCell();
+        }
```

Comment:
Iterating over the set of focusable elements in the entire table each time focus is changed is quite inefficient. Moreover you're limiting the focus ability only to certain elements `a, button, input, textarea, select, details, [tabindex]:not([tabindex="-1"]`. I'm wondering if it's a good idea..

What exactly this logic does? I see that this is being invoked specifically in the case when `TAB` is hit.

Reply:
The idea is to have the next behavior: jump over the cells with arrow keys, use f2 to go in/out of the cell, use tab to move between focusable elements in the table. We tried to reach the same behavior as the sapui5 grid table has.

`a, button, input, textarea, select, details, [tabindex]:not([tabindex="-1"]` are the selectors of all possible focusable elements. The opposite way is to bother users to apply the certain directive to focusable elements in the table cells.

Reply:
I see. Maybe we should then at least add a custom attribute that developers can use to specify the element is focusable?

Reply:
If user needs focusable `div` or something he can add `tabindex=0` to it, so it will be catched and added to the tabbable elements, because of the `[tabindex]:not([tabindex="-1"]` rule.

Reply:
About what you sent, they're also checking all that stuff but just in more progressive way and also checking iframe, video & audio elements because of their controls and so on.

So maybe we need to improve our approach, but as a bare minimum rules above is enough i think.